src/init.cpp

```diff
@@ -1326,11 +1333,14 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     if (onion_proxy.IsValid()) {
         SetProxy(NET_ONION, onion_proxy);
     } else {
-        if (args.IsArgSet("-onlynet") && IsReachable(NET_ONION)) {
+        // If -listenonion is set, then we will (try to) connect to the Tor control port
+        // later from the torcontrol thread and may retrieve the onion proxy from there.
+        const bool listenonion_disabled{!args.GetBoolArg("-listenonion", DEFAULT_LISTEN_ONION)};
```

Comment:
Since `DEFAULT_LISTEN_ONION=true`, this makes the InitError go away in some cases where it would be useful - i.e. in the (probably very typical) where a user just starts `./bitcoind -onlynet=onion` without providing any additional configuration and hopes everything works together out of the box. The user would have to actively set `-listenonion=0` or `listen=0` to still get this error.

I wonder if it might be more helpful to change the InitError to a info/warning (something along the lines of "Did you maybe forgot to provide a tor proxy? Disregard this you use torcontrol") that is not conditional on `-listenonion` Not a strong suggestion though, I am not sure either way.

Reply:
> this makes the InitError go away in some cases where it would be useful ... `./bitcoind -onlynet=onion`

That is exactly the case where we do not want an init error -- the problem described in https://github.com/bitcoin/bitcoin/issues/24980 which this PR aims to fix. Because it could as well work out of the box, giving an init error is too strict and is relaxed by this PR.

Now, it is true that we do not know for sure, at this stage, at startup, whether we would be able to connect to torcontrol and retrieve the proxy address from it. I think it is proper to allow startup and later, if the tor control thread fails to retrieve the proxy it will log the relevant error. That is similar (but not identical) to the case where `-proxy=` is given and at startup we do not know if it will work and actually proxy connections. Maybe later we would find out that nobody is listening on the given proxy address and will log an error then.

That is, we give init errors only if we are sure, guaranteed that it will not work.

> I wonder if it might be more helpful to change the InitError to a info/warning

I am a bit confused. Do you mean to keep the init error as it is in this PR, but in the case where we silently allow startup (the new relaxation in this PR) to log a warning? I think it would not be necessary because it could be annoying for the cases when all is ok. For the cases when it is not, then the tor control thread will log a proper message if/when the problem occurs. Also, "Disregard this if you use torcontrol" -- `-torcontrol` is irrelevant here -- even if not specified it has a reasonable default. The relevant setting is `-listenonion=1`.

Off topic rant: I find the name `listenonion` confusing, it should be something like `-auto_create_tor_listening_service_via_torcontrol` because `-listenonion` does not control whether or not we _listen_ on `127.0.0.1:8334` and the tor service could be setup in `tor.conf` (the "static" way in which we should have `-listenonion=0`, but would still be listening and accepting incoming tor connections).

Reply:
I was just thinking about the scenario where a user installs and starts tor, then `./bitcoind -onlynet=onion`, but didn't setup a Onion Service as described in `tor.md` because they didn't read the doc.
In that case they wouldn't make outbound connections and I think also get no error or warning from the torcontrol Thread unless they have `debug=tor` enabled.

Reply:
> but didn't setup a Onion Service as described in `tor.md`

There are two ways - [automatic](https://github.com/bitcoin/bitcoin/blame/2557429d2b82fe5ba308b940bcf14c3047634345/doc/tor.md#L70) and [manual](https://github.com/bitcoin/bitcoin/blame/2557429d2b82fe5ba308b940bcf14c3047634345/doc/tor.md#L162), which one do you mean?

What you describe above _could_ work with the automatic way:
1. `bitcoind` finds the tor control at `127.0.0.1:9051` (no need to specify `-torcontrol=127.0.0.1:9051` since that is the default).
2. The tor daemon tells `bitcoind` where is the tor cookie file and `bitcoind` is able to read it. Authentication is successful.
3. `bitcoind` automatically creates the onion service and starts listening and accepting incoming tor connections (locally at `127.0.0.1:8334`)
4. `bitcoind` retrieves the address of the socks5 tor proxy from tor control and sets `-onion` to it, for making outbound connections to tor peers.

If 2. fails then there will be some error in the log, either from [`TorController::auth_cb()`](https://github.com/bitcoin/bitcoin/blob/2557429d2b82fe5ba308b940bcf14c3047634345/src/torcontrol.cpp#L459) or from [`TorController::protocolinfo_cb()`](https://github.com/bitcoin/bitcoin/blob/2557429d2b82fe5ba308b940bcf14c3047634345/src/torcontrol.cpp#L562-L595).

If 4. fails, there will be an error from [`TorController::get_socks_cb()`](https://github.com/bitcoin/bitcoin/blob/2557429d2b82fe5ba308b940bcf14c3047634345/src/torcontrol.cpp#L337).

Reply:
I mean the automatic way.
If step 1) is not successful, there will be  repeated messages:
```
[tor] Error connecting to Tor control socket
[tor] Not connected to Tor control port 127.0.0.1:9051, trying to reconnect
```
that are usually hidden in the `tor` debug category, because with `-listenonion=1` being the default, we probably don't even intend to use anything but clearnet and don't need to see those. These messages would become kind of relevant though if we used `-onlynet=onion` because we'd not find any peers without being able to find tor control, so it might be helpful to log something unconditionally in this case.
But these are logging nits that don't have to be discussed in this PR and certainly shouldn't hold it off, so I think you should just go ahead and resolve this.

Reply:
> ... a user installs and starts tor, then ./bitcoind -onlynet=onion ...

For 1. to fail, then tor is not listening on `127.0.0.1:9051` which it does by default. So, maybe the user did not install and start tor or did so, but fiddled with its config and changed the listening address.

Anyway, maybe this change makes sense:

```diff
--- i/src/torcontrol.cpp
+++ w/src/torcontrol.cpp
@@ -612,13 +612,26 @@ void TorController::disconnected_cb(TorControlConnection& _conn)
     if (service.IsValid())
         RemoveLocal(service);
     service = CService();
     if (!reconnect)
         return;

-    LogPrint(BCLog::TOR, "Not connected to Tor control port %s, trying to reconnect\n", m_tor_control_
+    const auto onlynets = gArgs.GetArgs("-onlynet");
+    if (gArgs.IsArgSet("-onlynet") &&
+        std::any_of(onlynets.begin(), onlynets.end(), [](const auto& n) {
+            return ParseNetwork(n) == NET_ONION;
+        })) {
+        LogPrintf("Not connected to Tor control port %s, trying to reconnect. "
+                  "Use -listenonion=0 to disable the tor control thread.\n",
+                  m_tor_control_center);
+    } else {
+        LogPrint(BCLog::TOR,
+                 "Not connected to Tor control port %s, trying to reconnect. "
+                 "Use -listenonion=0 to disable the tor control thread.\n",
+                 m_tor_control_center);
+    }

     // Single-shot timer for reconnect. Use exponential backoff.
```

Keep it in the debug log only if `-onlynet=onion` is not given, because for users that did not install tor and do not care about tor and just run `bitcoind` this would be just noise.